feat: Add error marking to each question in each dimension group

teddy-f-47 opened the pull request:
This PR provides error marking to each individual question. Previously, when a dimension group contains an unfilled question, the border of the group is simply colored red without indicating which question is the issue. Now, in addition to coloring the border red, it will also make the unfilled question red.

I initially wanted to include the code changes here in PR #78, but decided against it so that PR #78 can be merged quickly. The code changes here probably require a bit more testing, even though I have tested them and they seem to work well. 😅 

Screenshots:
<img width="960" alt="ssFix1" src="https://user-images.githubusercontent.com/64476430/212145518-a29dab9f-6fec-4607-9eca-6a210628e06c.PNG">
<img width="960" alt="ssFix2" src="https://user-images.githubusercontent.com/64476430/212145522-3f5d077d-ce54-499f-9f74-ecf42995d020.PNG">
<img width="960" alt="ssFix3" src="https://user-images.githubusercontent.com/64476430/212145526-2fe2925c-696d-43f7-9270-5fab641dc346.PNG">
<img width="960" alt="ssFix4" src="https://user-images.githubusercontent.com/64476430/212145532-9e46ead2-fc5c-4588-81c7-1f6fb40b2faf.PNG">
<img width="702" alt="ssFix5" src="https://user-images.githubusercontent.com/64476430/212145533-920e9aae-b433-47bf-8c6e-42799d6cec17.PNG">
<img width="702" alt="ssFix6" src="https://user-images.githubusercontent.com/64476430/212145537-5a379098-f030-490f-9ecd-480d8cc912e7.PNG">

What's Changed:
 - frontend/components/tasks/affectiveAnnotation/inputs/Slider.vue: Add `error` prop to mark question with error with red text color
 - frontend/components/tasks/affectiveAnnotation/inputs/TextfieldWithSeq2Seq.vue: Add `error` prop to mark question with error with red text color
 - frontend/components/tasks/affectiveAnnotation/summary/SummaryInput.vue: Fill error prop, add `hasValidEntries` and `hasErrors` computed properties
 - frontend/components/tasks/affectiveAnnotation/emotions/EmotionsInput.vue: Fill error prop, add `hasValidEntries` and `hasErrors` computed properties
 - frontend/components/tasks/affectiveAnnotation/others/OthersInput.vue: Fill error prop, add `hasValidEntries` and `hasErrors` computed properties
 - frontend/components/tasks/affectiveAnnotation/offensive/OffensiveInput.vue: Make question text red in case of error, add `hasValidSubquestion3`, `hasValidSubquestion4`, and `hasErrors` computed properties
 - frontend/components/tasks/affectiveAnnotation/humor/HumorInput.vue: Make question text red in case of error, add `hasValidSubquestion3`, `hasValidSubquestion4`, and `hasErrors` computed properties
 - frontend/pages/projects/_id/affective-annotation/index.vue: Refactor validation mechanisms of SummaryInput, EmotionsInput, and OthersInput because the child components now have an internal validation property `hasValidEntries`

artikandri commented on frontend/components/tasks/affectiveAnnotation/others/OthersInput.vue:
```diff
@@ -271,6 +283,16 @@ export default {
+      return items.findIndex((item) => item === this.flagSliderUnclicked) === -1 && this.wishToAuthor.length > 0
+    },
+    hasErrors() {
+      console.log("others", (this.showErrors) ? !this.hasValidEntries : false)
```
Console.log?

teddy-f-47 replied:
Sorry for missing it 😅 removed XD

artikandri reviewed (commented):
Hello, I have checked the code and it seems OK other than one console.log accidentally pushed forward. I will come back here again after checked the functionalities xD

XeonBloomfield commented:
Thank you for separating #78 from this one. I agree that more testing is needed here, as erroneous error highlighting can strongly confuse annotators.

artikandri commented:
Hello, I have tested this branch. Generally it works fine, but it has problem with identifying the subquestion under the `humor` section (which probably means the same bug will happen with the `offensive` section as well since their code structures are the same XD   Can you check 

![image](https://user-images.githubusercontent.com/12537724/217219271-be5c9646-e091-4acc-b9c8-0f38ffc6d1c4.png)

Edit: 
Additionally, I think it's better to mark required questions as required (add * behind each question). And this is my preference, but I think it's better to add a new label which said "This question is required" rather than only highlighting it because it may pose a problem to people with color blindness, but I will leave this to your judgement.

teddy-f-47 commented:
> Hello, I have tested this branch. Generally it works fine, but it has problem with identifying the subquestion under the `humor` section (which probably means the same bug will happen with the `offensive` section as well since their code structures are the same XD Can you check
> 
> ![image](https://user-images.githubusercontent.com/12537724/217219271-be5c9646-e091-4acc-b9c8-0f38ffc6d1c4.png)
> 
> Edit: Additionally, I think it's better to mark required questions as required (add * behind each question). And this is my preference, but I think it's better to add a new label which said "This question is required" rather than only highlighting it because it may pose a problem to people with color blindness, but I will leave this to your judgement.

Hello, thank you for the feedback! I agree, only highlighting it is a problem for people with color blindness. It is an excellent point that I didn't think of 😅 . I will update it with an additional label as you said.

Also, with the subquestions under question 3 and question 4, I actually only implemented the highlighting mechanism on the main question itself and not on the subquestions. So, for example, if question 3 requires input, the question sentence will be red, but the marked subquestion will not be highlighted. Mainly because I found it very difficult to implement the checking mechanism for the subquestions 😅 . But now that I have a fresh mind to look at the problem, I think it can be done in some other way; I will see update the PR again with the subquestions error marking.

teddy-f-47 commented:
> Hello, I have tested this branch. Generally it works fine, but it has problem with identifying the subquestion under the `humor` section (which probably means the same bug will happen with the `offensive` section as well since their code structures are the same XD Can you check
> 
> ![image](https://user-images.githubusercontent.com/12537724/217219271-be5c9646-e091-4acc-b9c8-0f38ffc6d1c4.png)
> 
> Edit: Additionally, I think it's better to mark required questions as required (add * behind each question). And this is my preference, but I think it's better to add a new label which said "This question is required" rather than only highlighting it because it may pose a problem to people with color blindness, but I will leave this to your judgement.

Hello, I have updated the PR. Now the error message is added to the question, the bugs in Humor and Offensive sections have been solved, and the subquestion now also has the error message too :) Thank you for the suggestions!

haanh764 reviewed (approved):
Thank you for your great work! I've checked the code and tested the functions and everything seems oke for me. The new way to handle the required fields is better also! So I will give my approval XD

Btw, there's a pretty minor thing that in humor and offensive section, when the 2 first scales are selected, both question 3 and 4 has the red notification "Wymagany" ad here 
![image](https://user-images.githubusercontent.com/17596118/218147705-a9511edb-7fa9-4d13-926b-b352888025f4.png)
But, actually answering only 1 of them is enough to click Confirm. It could be a bit confuse to user but I think it's quite minor :sweat_smile:

teddy-f-47 commented:
> Btw, there's a pretty minor thing that in humor and offensive section, when the 2 first scales are selected, both question 3 and 4 has the red notification "Wymagany" ad here ![image](https://user-images.githubusercontent.com/17596118/218147705-a9511edb-7fa9-4d13-926b-b352888025f4.png) But, actually answering only 1 of them is enough to click Confirm. It could be a bit confuse to user but I think it's quite minor 😅

Thank you for the feedback :) Somehow I kept thinking that both questions need to be answered. But you are right. I have adjusted the code, so now answering only one of them is enough to remove the error label :)

haanh764 commented:
> > Btw, there's a pretty minor thing that in humor and offensive section, when the 2 first scales are selected, both question 3 and 4 has the red notification "Wymagany" ad here ![image](https://user-images.githubusercontent.com/17596118/218147705-a9511edb-7fa9-4d13-926b-b352888025f4.png) But, actually answering only 1 of them is enough to click Confirm. It could be a bit confuse to user but I think it's quite minor 😅
> 
> Thank you for the feedback :) Somehow I kept thinking that both questions need to be answered. But you are right. I have adjusted the code, so now answering only one of them is enough to remove the error label :)

yeah somehow I kept thinking like that also lol XD but when I tested the logic is different. Or maybe we can confirm it again about the logic XD

artikandri commented:
Hello, thank you for the fix and I'm sorry for the long reply. I have checked the code and tested it, and unfortunately I happened to find (?) a bug in which when we typed the answer in the modal and I have yet to click the send button, the required mark will be lost. I understand that this is most likely because I  used a `getter setter` instead of using 2 different fields to store the value, so can you fix that? (remove the `getter setter` and use two different fields with a combination of `watch` or `@change` event, there should be an example somewhere in the codebase). 

![image](https://user-images.githubusercontent.com/12537724/218765469-d4cc349e-2024-49e9-86b9-daf155c677c8.png)

teddy-f-47 commented:
> ....Unfortunately I happened to find (?) a bug in which when we typed the answer in the modal and I have yet to click the send button, the required mark will be lost. I understand that this is most likely because I used a `getter setter` instead of using 2 different fields to store the value, so can you fix that? (remove the `getter setter` and use two different fields with a combination of `watch` or `@change` event, there should be an example somewhere in the codebase).

Hello, I have fixed the bug. Not sure if I fixed it in the way you expected, but now the required mark won't be lost if the user hasn't submitted the answer:
<img width="694" alt="sssss" src="https://user-images.githubusercontent.com/64476430/218989110-19861490-21fd-4f74-91d5-f11c91bfb9bb.PNG">

artikandri commented on frontend/components/tasks/affectiveAnnotation/inputs/TextfieldModal.vue:
```diff
@@ -95,28 +95,21 @@ export default {
-        this.$emit('input', val)
-      }
+      dialogErrorMessage: '',
+      text: (this.value === this.emptyTextFlag) ? '' : this.value
```
The `data` field is only executed once before mounting the component. I haven't tested this, but from my understanding, this should return error on every tries after the 1st try, did you check? I think it's better to implement the `watch` block to watch the `value` and change the `text` value based on the `value`. There is an example somewhere in the code base

teddy-f-47 replied:
It actually appeared to work fine on multiple tests, but I understand your point. Basically `text` was only filled by `value` at the beginning before mounting, and after that they work independently from each other. Even though they basically have the same value all the time, it is quite prone to bugs, so I agree that using `watch` on `value` to change `text` is better. I have refactored it, thank you for the feedback :)

artikandri reviewed (commented):
Hello, thank you for the fix. I havent tested the code, it looks better but unless I misunderstood how Vue lifecycle works, it will return error, so can you check?

artikandri reviewed (approved):
Hello, thank you for the fix. I have read it and I think it looks fine, the only comment I have now is to remove `persistent` from `modal` as one of the users mentioned, but I think it should be put in separate task because we'll have to test how the required fields will behave with no `persistent`, so for now it's oke. Thank you, and approved.
